# 1.Import libraries and connect to Postgresql

In [14]:
!pip install psycopg2-binary sqlalchemy


In [16]:
!pip install ipython-sql

In [11]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [12]:

%sql postgresql://postgres:171004@localhost:5432/HumanResources_Prj

In [13]:
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

In [19]:
%sql SELECT * FROM hr.human_resources LIMIT 100 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
100 rows affected.


[('00-0037846', 'Kimmy', 'Walczynski', '06-04-91', 'Male', 'Hispanic or Latino', 'Engineering', 'Programmer Analyst I', 'Headquarters', '1/20/2002', None, 'Cleveland', 'Ohio'),
 ('00-0041533', 'Ignatius', 'Springett', '6/29/1984', 'Male', 'White', 'Business Development', 'Business Analyst', 'Headquarters', '04-08-19', None, 'Cleveland', 'Ohio'),
 ('00-0045747', 'Corbie', 'Bittlestone', '7/29/1989', 'Male', 'Black or African American', 'Sales', 'Solutions Engineer Manager', 'Headquarters', '10-12-10', None, 'Cleveland', 'Ohio'),
 ('00-0055274', 'Baxy', 'Matton', '9/14/1982', 'Female', 'White', 'Services', 'Service Tech', 'Headquarters', '04-10-05', None, 'Cleveland', 'Ohio'),
 ('00-0076100', 'Terrell', 'Suff', '04-11-94', 'Female', 'Two or More Races', 'Product Management', 'Business Analyst', 'Remote', '9/29/2010', '2029-10-29 06:09:38 UTC', 'Flint', 'Michigan'),
 ('00-0116166', 'Kacie', 'Offiler', '1/18/1971', 'Male', 'Asian', 'Engineering', 'Developer III', 'Headquarters', '09-01-18', None, 'Cleveland', 'Ohio'),
 ('00-0363185', 'Sandro', 'Admans', '11/19/1979', 'Male', 'Two or More Races', 'Product Management', 'Quality Engineer', 'Headquarters', '11-08-12', None, 'Cleveland', 'Ohio'),
 ('00-0380704', 'Eugene', 'Lehrahan', '10/14/1988', 'Female', 'Black or African American', 'Engineering', 'Developer I', 'Headquarters', '6/27/2007', None, 'Cleveland', 'Ohio'),
 ('00-0381660', 'Wainwright', 'Corfield', '12/13/1996', 'Male', 'Asian', 'Engineering', 'Business Systems Development Analyst', 'Headquarters', '2/20/2001', '2008-12-05 01:21:48 UTC', 'Cleveland', 'Ohio'),
 ('00-0419202', 'Dyann', 'Isoldi', '3/27/1980', 'Male', 'Two or More Races', 'Engineering', 'Web Developer I', 'Headquarters', '1/27/2005', None, 'Cleveland', 'Ohio'),
 ('00-0472287', 'Grantley', 'Oret', '09-06-75', 'Male', 'Two or More Races', 'Services', 'Service Tech II', 'Headquarters', '11-01-04', None, 'Cleveland', 'Ohio'),
 ('00-0472832', 'Elmore', 'Worner', '01-07-66', 'Female', 'White', 'Engineering', 'Business Systems Development Analyst', 'Headquarters', '12-05-00', None, 'Cleveland', 'Ohio'),
 ('00-0566380', 'Dud', 'Brain', '3/17/1984', 'Male', 'Two or More Races', 'Business Development', 'Business Analyst', 'Headquarters', '9/17/2008', None, 'Cleveland', 'Ohio'),
 ('00-0571075', 'Aguie', 'Conford', '11-02-71', 'Male', 'White', 'Business Development', 'Research Assistant II', 'Headquarters', '11/25/2015', None, 'Cleveland', 'Ohio'),
 ('00-0624189', 'Katerina', 'Rosborough', '8/20/1967', 'Male', 'Hispanic or Latino', 'Engineering', 'Analyst Programmer', 'Headquarters', '5/17/2019', None, 'Cleveland', 'Ohio'),
 ('00-0715212', 'Alida', 'Longley', '1/28/1973', 'Female', 'American Indian or Alaska Native', 'Accounting', 'Staff Accountant III', 'Headquarters', '02-04-02', None, 'Cleveland', 'Ohio'),
 ('00-0755645', 'Laraine', 'Petre', '05-11-67', 'Male', 'White', 'Engineering', 'Software Engineer I', 'Headquarters', '9/30/2010', None, 'Cleveland', 'Ohio'),
 ('00-0778934', 'Gareth', 'MacCook', '2/21/1987', 'Female', 'Black or African American', 'Legal', 'Senior Attorney', 'Remote', '2/18/2010', None, 'Flint', 'Michigan'),
 ('00-0794247', 'Scottie', 'Chestney', '2/23/1972', 'Female', 'Black or African American', 'Engineering', 'Data Visualization Specialist', 'Headquarters', '07-03-02', None, 'Cleveland', 'Ohio'),
 ('00-0948136', 'Christophorus', 'Boseley', '5/23/1983', 'Female', 'Two or More Races', 'Marketing', 'Senior Editor', 'Headquarters', '09-06-07', None, 'Cleveland', 'Ohio'),
 ('00-0971612', 'Arleyne', 'Froome', '08-01-99', 'Male', 'Two or More Races', 'Engineering', 'Software Consultant', 'Headquarters', '04-09-15', None, 'Cleveland', 'Ohio'),
 ('00-1051096', 'Todd', 'Cashen', '6/19/1999', 'Female', 'White', 'Accounting', 'Financial Analyst', 'Remote', '6/20/2014', None, 'Pittsburgh', 'Pennsylvania'),
 ('00-1052230', 'Elmo', 'McNee', '12/16/1988', 'Female', 'Black or African American', 'Services', 'Service Manager', 'Headquarters', '7/18/2002', '2006-05

# 2.Data cleaning 

In [20]:
%sql SET search_path TO hr;

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
Done.


[]

In [21]:
%%sql 
SELECT column_name, data_type
FROM information_schema.columns
WHERE table_schema = 'hr' AND table_name = 'human_resources';

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
13 rows affected.


[('id', 'text'),
 ('first_name', 'text'),
 ('last_name', 'text'),
 ('birthdate', 'text'),
 ('gender', 'text'),
 ('race', 'text'),
 ('department', 'text'),
 ('jobtitle', 'text'),
 ('location', 'text'),
 ('hire_date', 'text'),
 ('termdate', 'text'),
 ('location_city', 'text'),
 ('location_state', 'text')]

## Clean column 'birthdate' : 
- Ta cần chuyển hết sang định dạng 'y-m-d'
- Vấn đề : Có 2 kiểu format date cho cột này :   
    - Kiểu thứ nhất có dạng 'm/d/y' 
    - kiểu thứ hai có dạng 'm-d-y'  
    - Riêng với kiểu thứ hai , phần 'year' chỉ lấy 2 số cuối xy . Nếu xy > 25 (2025) , thì ta sẽ chuyển nó thành 19xy . Ngược lại , ta chuyển thành 20xy 

Cập nhật lần 1 : Đổi hết về format đã quy định

In [27]:
%%sql
update human_resources 
set birthdate = case 
  when birthdate like '%/%' then TO_CHAR(TO_DATE(birthdate, 'MM/DD/YYYY') , 'YYYY-MM-DD')
  when birthdate like '%-%' then TO_CHAR(TO_DATE(birthdate, 'MM-DD-YY'), 'YYYY-MM-DD')
  else birthdate 
end ;

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
22214 rows affected.


[]

Ép kiểu date

In [29]:

%%sql
alter table human_resources 
alter column birthdate type date
USING birthdate::date; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
Done.


[]

Cập nhật lần 2 : Trừ 100 năm cho những date có year > 2025 (ví dụ : 2066 -> 1966)

In [30]:
%%sql 
update human_resources
set birthdate = case 
  when extract(year from birthdate) > 2025 then birthdate - interval '100 years'
  else birthdate
end;

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
22214 rows affected.


[]

Xem kết quả 

In [39]:
%%sql
SELECT birthdate FROM human_resources LIMIT 10

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
10 rows affected.


[(datetime.date(1971, 10, 19),),
 (datetime.date(1998, 10, 31),),
 (datetime.date(1991, 3, 6),),
 (datetime.date(1988, 2, 9),),
 (datetime.date(1971, 2, 13),),
 (datetime.date(1989, 3, 1),),
 (datetime.date(1998, 11, 9),),
 (datetime.date(1994, 6, 26),),
 (datetime.date(1995, 1, 21),),
 (datetime.date(1986, 10, 29),)]

## Clean column 'hire_date': Xử lý tương tự như cột 'birthdate' 

Cập nhật lần 1 : Đổi hết sang format quy định 

In [35]:
%%sql 
update human_resources 
set hire_date = case 
  when hire_date like '%/%' then to_char(to_date(hire_date , 'MM/DD/YYYY') , 'YYYY-MM-DD')
  when hire_date like '%-%' then to_char(to_date(hire_date , 'MM/DD/YY') , 'YYYY-MM-DD')
end ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
22214 rows affected.


[]

Ép kiểu date 

In [36]:
%%sql 
alter table human_resources 
alter column hire_date type date 
using hire_date::date ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
Done.


[]

Cập nhật lần 2 : Trừ 100 năm cho những date có year > 2025 (ví dụ : 2066 -> 1966)

In [37]:
%%sql 
update human_resources 
set hire_date = case 
  when extract(year from hire_date) > 2025 then hire_date - interval '100 years'
  else hire_date
end; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
22214 rows affected.


[]

Kiểm tra 

In [38]:
%%sql
select hire_date from human_resources
limit 10 ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
10 rows affected.


[(datetime.date(2013, 7, 28),),
 (datetime.date(2016, 4, 12),),
 (datetime.date(2012, 3, 6),),
 (datetime.date(2011, 11, 5),),
 (datetime.date(2003, 3, 12),),
 (datetime.date(2005, 6, 11),),
 (datetime.date(2011, 9, 29),),
 (datetime.date(2010, 1, 12),),
 (datetime.date(2008, 12, 2),),
 (datetime.date(2009, 7, 28),)]

## Clean column 'termdate'

In [42]:
%sql select termdate from human_resources limit 10 ;

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
10 rows affected.


[(None,),
 (None,),
 (None,),
 (None,),
 (None,),
 (None,),
 (None,),
 (None,),
 (None,),
 ('2011-06-25 21:03:08 UTC',)]

Chuyển từ dạng timestamp sang format quy định 

In [43]:
%%sql 
update human_resources 
set termdate = to_char(to_timestamp(termdate, 'YYYY-MM-DD HH24:MI:SS UTC') , 'YYYY-MM-DD')
where termdate is not null and termdate != ' '; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
3929 rows affected.


[]

Ép lại kiểu 

In [44]:
%%sql 
alter table human_resources 
alter column termdate type date 
using termdate::date ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
Done.


[]

Kiểm tra 

In [49]:
%sql select termdate from human_resources limit 10 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
10 rows affected.


[(datetime.date(2017, 7, 28),),
 (datetime.date(2030, 5, 12),),
 (datetime.date(2029, 10, 29),),
 (datetime.date(2008, 12, 5),),
 (datetime.date(2006, 5, 22),),
 (datetime.date(2022, 7, 12),),
 (datetime.date(2022, 2, 25),),
 (datetime.date(2030, 3, 21),),
 (datetime.date(2024, 10, 9),),
 (datetime.date(2007, 10, 31),)]

# 3.QUERIES 

## 1. What is the gender breakdown of employees in the company?

In [50]:
%%sql 
select count(id) as number_of_employees , gender
from human_resources 
group by gender ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
3 rows affected.


[(11288, 'Male'), (10321, 'Female'), (605, 'Non-Conforming')]

## 2. What is the race/ethnicity breakdown of employees in the company?

In [52]:
%%sql 
select count(id) as number_of_employees , race 
from human_resources 
group by race
order by count(id) desc ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
7 rows affected.


[(6328, 'White'),
 (3648, 'Two or More Races'),
 (3619, 'Black or African American'),
 (3562, 'Asian'),
 (2501, 'Hispanic or Latino'),
 (1327, 'American Indian or Alaska Native'),
 (1229, 'Native Hawaiian or Other Pacific Islander')]

## 3. What is the age distribution of employees in the company?

Đầu tiên , tạo thêm cột 'age'

In [53]:
%%sql 
alter table human_resources
add column age integer ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
Done.


[]

In [54]:
%%sql 
update human_resources 
set age = extract(year from age(now() , birthdate));

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
22214 rows affected.


[]

Kế tiếp , thực hiện query

In [55]:
%%sql 
select concat(floor(age/10)* 10 , '-' , floor(age/10)*10 + 9)as age_range , count(id) as number_of_employees
from human_resources 
group by age_range 
order by age_range asc ;

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
4 rows affected.


[('20-29', 4278), ('30-39', 6165), ('40-49', 5886), ('50-59', 5885)]

## 4. How many employees work at headquarters versus remote locations?

In [56]:
%%sql 
select location ,  count(id) as number_of_employees 
from human_resources 
group by location  ;

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
2 rows affected.


[('Headquarters', 16715), ('Remote', 5499)]

## 5. What is the average length of employment for employees who have been terminated?

In [60]:
%%sql 
select round( avg( extract (year from age(termdate, hire_date))) , 2) as average_length_of_employment_years
from human_resources 
where termdate is not null and termdate < now() ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
1 rows affected.


[(Decimal('7.68'),)]

## 6. How does the gender distribution vary across departments?

In [61]:
%%sql 
select department , gender , count(id) as number_of_employees
from human_resources 
group by  department , gender
order by department asc ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
38 rows affected.


[('Accounting', 'Male', 1711),
 ('Accounting', 'Non-Conforming', 91),
 ('Accounting', 'Female', 1531),
 ('Auditing', 'Female', 24),
 ('Auditing', 'Male', 28),
 ('Business Development', 'Female', 757),
 ('Business Development', 'Male', 836),
 ('Business Development', 'Non-Conforming', 49),
 ('Engineering', 'Female', 3120),
 ('Engineering', 'Non-Conforming', 193),
 ('Engineering', 'Male', 3373),
 ('Human Resources', 'Female', 861),
 ('Human Resources', 'Male', 904),
 ('Human Resources', 'Non-Conforming', 42),
 ('Legal', 'Female', 140),
 ('Legal', 'Male', 162),
 ('Legal', 'Non-Conforming', 9),
 ('Marketing', 'Non-Conforming', 5),
 ('Marketing', 'Male', 256),
 ('Marketing', 'Female', 233),
 ('Product Management', 'Non-Conforming', 15),
 ('Product Management', 'Female', 277),
 ('Product Management', 'Male', 349),
 ('Research and Development', 'Male', 531),
 ('Research and Development', 'Female', 513),
 ('Research and Development', 'Non-Conforming', 40),
 ('Sales', 'Non-Conforming', 47),
 ('Sales', 'Male', 946),
 ('Sales', 'Female', 839),
 ('Services', 'Non-Conforming', 33),
 ('Services', 'Male', 853),
 ('Services', 'Female', 800),
 ('Support', 'Female', 437),
 ('Support', 'Non-Conforming', 36),
 ('Support', 'Male', 481),
 ('Training', 'Non-Conforming', 45),
 ('Training', 'Female', 789),
 ('Training', 'Male', 858)]

## 7. What is the distribution of job titles across the company?

In [69]:
%%sql 
select jobtitle , count(id)as number_of_employees 
from human_resources 
group by jobtitle 
order by number_of_employees  desc ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
185 rows affected.


[('Research Assistant II', 754),
 ('Business Analyst', 708),
 ('Human Resources Analyst II', 613),
 ('Research Assistant I', 538),
 ('Account Executive', 505),
 ('Data Visualization Specialist', 457),
 ('Staff Accountant I', 441),
 ('Human Resources Analyst', 408),
 ('Software Engineer I', 397),
 ('Systems Administrator I', 374),
 ('Analyst Programmer', 372),
 ('Senior Developer', 361),
 ('Data Coordiator', 356),
 ('Business Systems Development Analyst', 341),
 ('Project Manager', 340),
 ('Service Tech III', 339),
 ('Software Consultant', 336),
 ('Help Desk Technician', 336),
 ('Service Manager', 327),
 ('Service Tech', 327),
 ('Service Tech II', 327),
 ('Service Coordinator', 327),
 ('Desktop Support Technician', 321),
 ('Senior Cost Accountant', 317),
 ('Cost Accountant', 308),
 ('Structural Engineer', 307),
 ('Tax Accountant', 303),
 ('Recruiter', 234),
 ('Operator', 225),
 ('Customer Success Manager', 214),
 ('Account Manager', 214),
 ('Pre-Sales Consultant', 214),
 ('Solutions Engineer', 214),
 ('Relationshiop Manager', 214),
 ('Solutions Engineer Manager', 214),
 ('Research Associate', 208),
 ('Senior Recruiter', 204),
 ('HR Manager', 204),
 ('Administrative Assistant I', 196),
 ('Trainer I', 182),
 ('Trainer II', 182),
 ('Senior Trainer', 182),
 ('Trainer III', 182),
 ('Assistant Trainer', 182),
 ('Training Manager', 182),
 ('Budget/Accounting Analyst IV', 181),
 ('Junior Trainer', 181),
 ('Budget/Accounting Analyst II', 166),
 ('Budget/Accounting Analyst I', 161),
 ('Budget/Accounting Analyst III', 160),
 ('Senior Quality Engineer', 152),
 ('Senior Editor', 141),
 ('Quality Control Specialist', 138),
 ('Media Manager I', 129),
 ('Support Staff', 104),
 ('Staff Accountant IV', 102),
 ('Software Engineer III', 100),
 ('Web Developer I', 99),
 ('Staff Accountant III', 96),
 ('Accounting Assistant IV', 95),
 ('Programmer Analyst I', 94),
 ('Software Engineer IV', 94),
 ('Programmer Analyst II', 93),
 ('Editor', 92),
 ('Chief Design Engineer', 92),
 ('Help Desk Operator', 92),
 ('Computer Systems Analyst II', 92),
 ('Database Administrator II', 91),
 ('Computer Systems Analyst I', 90),
 ('Accounting Assistant I', 89),
 ('Accountant II', 89),
 ('Software Test Engineer IV', 88),
 ('Software Test Engineer I', 88),
 ('Systems Administrator IV', 88),
 ('Accountant IV', 87),
 ('Software Test Engineer III', 87),
 ('Programmer I', 86),
 ('Accountant III', 86),
 ('Software Engineer II', 84),
 ('Database Administrator I', 84),
 ('Systems Administrator III', 84),
 ('Database Administrator IV', 83),
 ('Database Administrator III', 83),
 ('Developer III', 83),
 ('Accounting Assistant II', 83),
 ('Developer I', 81),
 ('Programmer III', 81),
 ('Programmer Analyst III', 80),
 ('Staff Accountant II', 80),
 ('Programmer II', 79),
 ('Accountant I', 79),
 ('Web Developer II', 77),
 ('VP Accounting', 77),
 ('Computer Systems Analyst III', 76),
 ('Developer IV', 76),
 ('Programmer Analyst IV', 76),
 ('Software Test Engineer II', 74),
 ('Systems Administrator II', 74),
 ('Accounting Assistant III', 72),
 ('Computer Systems Analyst IV', 72),
 ('Web Developer III', 71),
 ('Financial Analyst', 68),
 ('Developer II', 67),
 ('Web Developer IV', 66),
 ('Paralegal', 64),
 ('Programmer IV', 63),
 ('Human Resources Manager', 63),
 ('Quality Engineer', 62),
 ('Senior Financial Analyst', 59),
 ('Support Staff II', 59),
 ('Administrative Officer', 59),
 ('Product Engineer', 59),
 ('Content Developer II', 54),
 ('Design Engineer', 54),
 ('Content Developer', 54),
 ('Content Developer III', 54),
 ('Internal Auditor', 52),
 ('Media Manager II', 49),
 ('Analog Circuit Design manager', 48),
 ('Structural Analysis Engineer', 48),
 ('Administrative Assistant II', 46),
 ('Assistant Manager', 45),
 ('VP Quality Control', 43),
 ('VP Marketing', 42),
 ('VP Product Management', 41),
 ('Administrative Assistant', 40),
 ('Research Assistant', 40),
 ('Executive Assistant', 40),
 ('Senior Attorney', 40),
 ('Attorney', 40),
 ('Information Systems Manager', 35),
 ('Web Design

## 8. Which department has the highest turnover rate?

"Turnover rate" hay "Tỷ lệ luân chuyển" thường đề cập đến tỷ lệ nhân viên rời khỏi một công ty hoặc phòng ban và cần được thay thế. Tỷ lệ này có thể được tính bằng số lượng nhân viên nghỉ việc trong một khoảng thời gian nhất định chia cho số lượng nhân viên trung bình trong công ty hoặc phòng ban đó trong cùng khoảng thời gian đó.d.

In [63]:
%%sql 
select department , count(id) as total_employees , 
  sum(case when termdate is null then 1 else 0 end ) as total_active_employees ,
  sum(case when  termdate is not null and termdate < now() then 1 else 0 end ) as total_terminated_employees ,
  round( (sum(case when  termdate is not null and termdate < now() then 1 else 0 end )::decimal /count(id))  , 3 ) as turnover_rate
from human_resources 
group by department 
order by turnover_rate desc ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
13 rows affected.


[('Auditing', 52, 40, 10, Decimal('0.192')),
 ('Legal', 311, 248, 47, Decimal('0.151')),
 ('Training', 1692, 1401, 217, Decimal('0.128')),
 ('Engineering', 6686, 5501, 837, Decimal('0.125')),
 ('Support', 954, 772, 119, Decimal('0.125')),
 ('Human Resources', 1807, 1498, 225, Decimal('0.125')),
 ('Research and Development', 1084, 872, 134, Decimal('0.124')),
 ('Sales', 1832, 1497, 228, Decimal('0.124')),
 ('Accounting', 3333, 2747, 403, Decimal('0.121')),
 ('Services', 1686, 1393, 204, Decimal('0.121')),
 ('Product Management', 641, 527, 75, Decimal('0.117')),
 ('Business Development', 1642, 1367, 181, Decimal('0.110')),
 ('Marketing', 494, 422, 53, Decimal('0.107'))]

-> Auditing là phòng ban có turnover rate cao nhất 

## 9. What is the distribution of employees across locations by state?

In [65]:
%%sql 
select location_state , count(id) as number_of_employees 
from human_resources 
group by location_state 
order by count(id) desc ;

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
7 rows affected.


[('Ohio', 18025),
 ('Pennsylvania', 1115),
 ('Illinois', 868),
 ('Indiana', 700),
 ('Michigan', 673),
 ('Kentucky', 451),
 ('Wisconsin', 382)]

## 10. What is the net change over years between hired_employee's and terminated_employee's counts  

In [66]:
%%sql 
select 
    years, 
    hires, 
    terminations, 
    (hires - terminations) AS net_change,
    ROUND(((hires - terminations) / hires::float * 100)::numeric, 2) AS net_change_percent
from (
    select 
         extract(year from hire_date ) AS years, 
        count(*) as hires, 
        SUM(CASE WHEN termdate is not null AND termdate <= now() THEN 1 ELSE 0 END) AS terminations
    from 
        human_resources
    group by 
        years
) subquery
order by years asc;

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
21 rows affected.


[(Decimal('2000'), 220, 31, 189, Decimal('85.91')),
 (Decimal('2001'), 1122, 203, 919, Decimal('81.91')),
 (Decimal('2002'), 1067, 174, 893, Decimal('83.69')),
 (Decimal('2003'), 1142, 203, 939, Decimal('82.22')),
 (Decimal('2004'), 1135, 211, 924, Decimal('81.41')),
 (Decimal('2005'), 1097, 201, 896, Decimal('81.68')),
 (Decimal('2006'), 1118, 199, 919, Decimal('82.20')),
 (Decimal('2007'), 1090, 161, 929, Decimal('85.23')),
 (Decimal('2008'), 1108, 153, 955, Decimal('86.19')),
 (Decimal('2009'), 1140, 164, 976, Decimal('85.61')),
 (Decimal('2010'), 1099, 147, 952, Decimal('86.62')),
 (Decimal('2011'), 1101, 127, 974, Decimal('88.47')),
 (Decimal('2012'), 1103, 128, 975, Decimal('88.40')),
 (Decimal('2013'), 1105, 109, 996, Decimal('90.14')),
 (Decimal('2014'), 1053, 101, 952, Decimal('90.41')),
 (Decimal('2015'), 1059, 99, 960, Decimal('90.65')),
 (Decimal('2016'), 1122, 84, 1038, Decimal('92.51')),
 (Decimal('2017'), 1091, 70, 1021, Decimal('93.58')),
 (Decimal('2018'), 1147, 62, 1085, Decimal('94.59')),
 (Decimal('2019'), 1083, 62, 1021, Decimal('94.28')),
 (Decimal('2020'), 1012, 44, 968, Decimal('95.65'))]

## 11. What is the average tenure distribution for each department?

In [68]:
%%sql 
select department , round(  avg( (termdate - hire_date)::float /365 )::numeric  , 3) as avg_tenure 
from human_resources 
where termdate is not null 
group by department 
order by avg_tenure desc ; 

 * postgresql://postgres:***@localhost:5432/HumanResources_Prj
13 rows affected.


[('Sales', Decimal('11.122')),
 ('Research and Development', Decimal('10.903')),
 ('Services', Decimal('10.618')),
 ('Engineering', Decimal('10.578')),
 ('Support', Decimal('10.556')),
 ('Business Development', Decimal('10.501')),
 ('Marketing', Decimal('10.394')),
 ('Accounting', Decimal('10.354')),
 ('Human Resources', Decimal('9.983')),
 ('Product Management', Decimal('9.948')),
 ('Training', Decimal('9.690')),
 ('Auditing', Decimal('9.612')),
 ('Legal', Decimal('9.362'))]

# 4.Summary of findings : 
1. Số lượng nhân viên nam nhiều hơn nữ , tuy nhiên chênh lệch không quá nhiều (11288 vs 10321)
2. Người da trắng chiếm đa số trong công ty . Ngược lại người Hawaii bản địa và nhóm thuộc các đảo Thái Bình Dương có số lượng ít nhất 
3. Độ tuổi chiếm đa số trong công ty là từ 30 tới 39 tuổi 
4. Số lượng làm việc trực tiếp tại trụ sở cao hơn khoảng 3 lần so với số làm việc từ xa (remote)
5. Độ dài trung bình số năm làm việc của nhân viên là khoảng 7,68 năm 
6. Sự phân bổ giới tính giữa nam và nữ ở các phòng ban khác nhau là khá đồng đều , nam giới có xu hướng nhiều hơn đôi chút 
7. Các công việc ở vị trí trợ lý (Assistant) hoặc phân tích (Analyst) có số lượng nhân viên đông nhất 
8. Phòng ban có tỉ lệ luân chuyển (turnover rate) cao nhất là Kiểm Toán (Auditing) . Ngược lại , thấp nhất là Marketing 
9. Bang Ohio là bang có nhiều nhân viên nhất với 18025 , bỏ xa bang thứ nhì là Pennsylvania (1115) . Wisconsin là bang có ít nhân viên nhất (382) 
10. Tỉ lệ chênh lệch (net change) giữa số lượng nhân viên được tuyển mới và số lượng nghỉ việc , nhìn chung là tăng đều qua các năm . Có một khoảng giảm nhẹ từ 2000 tới 2005 , nhưng sau đó đã tăng đều trở lại 
11. Nhiệm kỳ làm việc trung bình ở các phòng ban hầu hết đều trên 9 năm . Trong đó , dài nhất là phòng ban Sales với khoảng trên 11 năm 